### DISCUSSION TITLE
Next.js APIs should be auto-documentable and exist in a "routes" folder rather than a "pages" folder

### DISCUSSION THREAD

**Discussion by miloofcroton** [vercel/next.js#36767] [Ideas]
There are two issues that jump out at me: one feature that is lacking for professional API development and one semantic in nature.

### OpenAPI specs cannot be generated automatically

There is only one library that seems to work with Next.js for generating OpenAPI specs: next-swagger-doc. It is not an automatic tool. All specs are manually defined via JSDoc, which means there are multiple sources of truth and no real contract between the spec and the implementation.

There is a library I prefer when using Express, and it is called express-oas-generator. What it does is spy on all of your server's routes, and it generates OpenAPI specs based on what the server responds with. It cannot truly parse the code itself, but its spying can be extremely useful when paired with thorough testing. E2E testing therefore creates this contract between the code and the spec. If you test every response type, you will get good spec coverage.

I think this is something that Next.js could handle in the framework. But if not, it's something that is lacking in the community.

### /pages -> /routes

Here's what always annoyed me about `pages`: what if I want to deploy just a backend? I still have to write something at `/pages/api/someEndpoint.ts`. This is just nonsensical. I would rather it be called `routes`, which is an isomorphic concept. A "page" is only really a front end concept. Do you want Nextjs's API route functionality to _only_ be used as a backend-for-frontend in the same app that you have your front end, or do you want people to look at it as potentially a general purpose tool for writing backends quickly and painlessly?

I propose `/routes` and `/src/routes` in addition to or in replacement to `/pages/` and `/src/pages`. I would be fine with this as a hardcoded value, but it also seems reasonable to have literally a single config value that can be defined for where your `pages` dir is.

Also note that people use the 'router' abstraction on the front end and back end, so it _is_ a universal concept. There's react router, and there's express router.

(_Originally posted by @miloofcroton in https://github.com/vercel/next.js/discussions/23640#discussioncomment-2705487_)

**Comment by GabenGar** [comment]
NextJS is a frontend server effectively, so using it for a backend-only is kinda strange, seeing as its main selling point over a typical persistent server is server-rendered pages. Otherwise it's pretty limited due to being serverless and therefore not fit for running background tasks outside of request/response cycle.

**Reply by miloofcroton** [reply]
Well, you only half answered the topic at hand, because even in a full stack setting, you are putting /api inside of /pages. That just logically doesn't make sense.

It does useful things in an API setting though. Why reject that usecase? It handles plenty of general API development. One case is where you lean towards an RPC style of development, so that you don't care about the HTTP verb.

Another situation is where you might start the app as a monolithic full stack app, and then you later want to split them into multiple packages. You will probably want to keep using Next for the front and back end in this case, so again you're in a situation where you want `/routes` as a more useful term for what the back end is doing.

**Reply by miloofcroton** [reply]
Just thought of something else. This might be a more annoying fix, as it could break some code? I don't know.

But you could make all .jsx and .tsx files under /routes compile as React components and all .ts and .js files compile as API routes. This would make the back end a little more flexible once again.

**Reply by GabenGar** [reply]
> you are putting /api inside of /pages. That just logically doesn't make sense.

As far as client-side navigation concerned page components also double up as API routes.

> and all .ts and .js files compile as API routes. 

More like .js, .cjs, .mjs, .ts, .cts and .mts files

Another problem is `/routes` doesn't imply the API routes specifically. `Routes` on the classical backend is a collection of endpoints bundled by some business logic and they can return absolutely anything, not just JSONs. So separating them by `/routes` and `/pages` isn't logical either, on top of breaking file-system based routing convention.
